aws/resource_aws_msk_scram_secret.go

```diff
@@ -0,0 +1,216 @@
+	}
+	d.SetId(d.Get("cluster_arn").(string))
+
+	if len(out.UnprocessedScramSecrets) != 0 {
+		return fmt.Errorf("there were unprocessed secrets during association: %s", out.UnprocessedScramSecrets)
+	}
+
+	return resourceAwsMskScramSecretRead(d, meta)
+}
+
+func resourceAwsMskScramSecretRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kafkaconn
+
+	scramSecrets, err := readMSKClusterSecrets(conn, d.Get("cluster_arn").(string))
+	if err != nil {
+		return fmt.Errorf("failed lookup secrets %s", err)
+	}
+
+	allSecrets := filterMskExistingSecrets(expandStringList(d.Get("secret_arn_list").([]interface{})), scramSecrets)
+
+	d.SetId(d.Get("cluster_arn").(string))
+	d.Set("arn", d.Get("cluster_arn").(string))
+	d.Set("scram_secrets", aws.StringValueSlice(allSecrets))
+
+	return nil
+}
+
+func resourceAwsMskScramSecretUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kafkaconn
+
+	existingSecrets := expandStringList(d.Get("scram_secrets").([]interface{}))
+
+	updateSecrets, deleteSecrets := filterMskSecretsForDeletion(expandStringList(d.Get("secret_arn_list").([]interface{})), existingSecrets)
+
+	out, err := associateMSKClusterSecrets(conn, d.Get("cluster_arn").(string), updateSecrets)
+	if err != nil {
+		return fmt.Errorf("error associating credentials with MSK cluster: %s", err)
+	}
+
+	if len(out.UnprocessedScramSecrets) != 0 {
+		return fmt.Errorf("there were unprocessed secrets during association: %s", out.UnprocessedScramSecrets)
+	}
+
+	if len(deleteSecrets) > 0 {
+		deleteOutput, err := conn.BatchDisassociateScramSecret(&kafka.BatchDisassociateScramSecretInput{
+			ClusterArn:    aws.String(d.Get("cluster_arn").(string)),
+			SecretArnList: deleteSecrets,
+		})
+		if err != nil {
+			return fmt.Errorf("error disassociating credentials with MSK cluster: %s", err)
+		}
+
+		if len(deleteOutput.UnprocessedScramSecrets) != 0 {
+			return fmt.Errorf("there were unprocessed secrets during association: %s", out.UnprocessedScramSecrets)
+		}
+	}
+
+	return resourceAwsMskScramSecretRead(d, meta)
+}
+
+func resourceAwsMskScramSecretDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kafkaconn
+
+	_, err := conn.BatchDisassociateScramSecret(&kafka.BatchDisassociateScramSecretInput{
+		ClusterArn:    aws.String(d.Get("cluster_arn").(string)),
+		SecretArnList: expandStringList(d.Get("secret_arn_list").([]interface{})),
+	})
+	if err != nil {
+		return fmt.Errorf("error disassociating credentials with MSK cluster: %s", err)
+	}
+
+	return nil
+}
+
+func readMSKClusterSecrets(conn *kafka.Kafka, clusterArn string) ([]*string, error) {
+	input := &kafka.ListScramSecretsInput{
+		ClusterArn: aws.String(clusterArn),
+	}
+
+	var scramSecrets []*string
+	err := conn.ListScramSecretsPages(input,
+		func(page *kafka.ListScramSecretsOutput, lastPage bool) bool {
+			scramSecrets = append(scramSecrets, page.SecretArnList...)
```

Comment:
to err on the side of caution and avoid unwanted panics:
```suggestion
			if page == nil {
			    return !lastPage
			}
			scramSecrets = append(scramSecrets, page.SecretArnList...)
```